### PR TITLE
Auto-save support for operator mode

### PR DESCRIPTION
**Pull request by habdelra**
This PR introduces auto-save to operator mode. After 5 seconds of inactivity to a card (activity being a change to any field), the card will be auto saved. The notable updates in this PR include:
- introduced a card instance subscription mechanism to the card runtime. in the card-api module scope I have added a WeakMap that holds the subscriptions for all the card instances. A subscriber is a synchronous callback that receives the field name that has changed and the new value. In the setter for the card instances (as well as the `WatchedArray` for the plural fields) the subscriber callbacks are invoked when a field value is changed. (additionally an unsubscribe is also included--the subscribe and unsubscribe should always be paired)
- updated the `StackItem` component so that on construction it subscribes to changes for its card instance, and upon destruction it unsubscribes to changes for its card instance. additionally, using ember concurrency, this component will invoke it's `@save()` argument 5 seconds after the last change to the card instance
- update the card service deserialization such that for newly created card instances that have not yet been saved, after the save occurs we mutate the instance with the updates from the server instead of creating a whole new instance. this disconnect was breaking object equality for newly created cards instances after they were saved
- updated linksTo and linksToMany deserialization such that if we have already loaded a field for the instance we'll reuse that field value the next time we deserialize the instance (for updates--this goes back to the previous point where newly created card instances were not triple equal to themselves after they were saved).
- in tests if you tear down the test while the save is still in flight (because auto save can happen at any time--we tune the auto save to 0ms after the instance changes in tests), you'll experience ember concurrency errors around cancelling a perform. To work around this, we have added a `[data-test-save-idle]` attribute to the operator mode container component. I have included a `waitFor('[data-test-save-idle]')` in the `afterEach()` for the operator mode integration tests to prevent these errors. This change does mean, though, that in general after setting a field in operator mode, you'll want to follow that with a `waitFor` to wait for the change to be saved so that the underlying data is consistent for the tests.

Peek 2023-08-01 12-24

### CODE REVIEW

**Review by burieberry — COMMENTED** [review]
Tested this feature by editing a nested `contained` card that opened up in the stack. (I'm not sure what the plans are for opening contained cards in the stack, so this comment might be irrelevant.) The first edit saved as expected. However, the follow up edits on the same card did not trigger a save. I tried the same actions on a regular card and did not see any problems there.

**Comment by habdelra** [comment]
> Tested this feature by editing a nested `contained` card that opened up in the stack. (I'm not sure what the plans are for opening contained cards in the stack, so this comment might be irrelevant.) The first edit saved as expected. However, the follow up edits on the same card did not trigger a save. I tried the same actions on a regular card and did not see any problems there.

yeah that was the issue i mentioned this morning in standup that i'm working on, i think it's fixed now. but fundamentally we have a huge pile of issues in the `container.gts` where we have async actions and we are casting EC tasks to promises. these are very problematic in that the component could be destroyed during async, and the ECtask casting to a promise means that we can't have cancellable tasks (we see this in the this.write in the container--it can't be cancelled because it is awaited in an async action). there should probably be a followup refactor to clean up the async issues in the `container.gts` (I added a bunch of TODO's for the async issues that I saw).

**Review by burieberry — APPROVED** [review]
Works as described

File: packages/host/app/components/operator-mode/container.gts [hunk 1]
```diff
@@ -199,6 +199,8 @@ export default class OperatorModeContainer extends Component<Signature> {
     }
   }
 
+  // TODO probably close needs to save--not revert, since with auto save
+  // it's not obvious what you are reverting to
```
Comment (habdelra) [inline packages/host/app/components/operator-mode/container.gts]: we just discussed in planning and yes--close should save

**Review by jurgenwerk — COMMENTED** [review]
Code looks good to me. I really appreciate all the comments! 

I noticed that when a card change gets saved and you close the stack items down to the index card, the items in the index card will reload and the newly updated card will get reordered to the end of the list which is unexpected. But I think that might be unrelated to this PR - maybe we can create a ticket for that.

Another thing is that when I change an input to in invalid value, the mechanism will still save the data. This is confusing - did it save invalid data?

image

**Comment by habdelra** [comment]
> Code looks good to me. I really appreciate all the comments!
> 
> I noticed that when a card change gets saved and you close the stack items down to the index card, the items in the index card will reload and the newly updated card will get reordered to the end of the list which is unexpected. But I think that might be unrelated to this PR - maybe we can create a ticket for that.
> 
> Another thing is that when I change an input to in invalid value, the mechanism will still save the data. This is confusing - did it save invalid data?
> 
> image

cool, i'll check out both these issues (and create a ticket if need be). I think for the invalid data scenario, even if it was being saved like that in the past, we should definitely not do that anymore. Perhaps we can add an indicator as well to the title bar that auto saving has been interrupted due to invalid data

**Comment by tintinthong** [comment]
> Code looks good to me. I really appreciate all the comments!
> 
> I noticed that when a card change gets saved and you close the stack items down to the index card, the items in the index card will reload and the newly updated card will get reordered to the end of the list which is unexpected. But I think that might be unrelated to this PR - maybe we can create a ticket for that.
> 
> Another thing is that when I change an input to in invalid value, the mechanism will still save the data. This is confusing - did it save invalid data?
> 
> image

the checks for input validity at least the way its designed currently doesn't enforce this. Invalid inputs can be saved but you'll notice because their not deserializable they will be saved as "null" in the json files. Most fields/cards which do this tap in to the deserialize hook

**Comment by habdelra** [comment]
> > Code looks good to me. I really appreciate all the comments!
> > I noticed that when a card change gets saved and you close the stack items down to the index card, the items in the index card will reload and the newly updated card will get reordered to the end of the list which is unexpected. But I think that might be unrelated to this PR - maybe we can create a ticket for that.
> > Another thing is that when I change an input to in invalid value, the mechanism will still save the data. This is confusing - did it save invalid data?
> > image
> 
> the checks for input validity at least the way its designed currently doesn't enforce this. Invalid inputs can be saved but you'll notice because their not deserializable they will be saved as "null" in the json files. Most fields/cards which do this tap in to the deserialize hook

ahhh, i see. so it looks like we don't have an official API for field validation--but rather the ethereum address card has some of it's own special logic to deal with validation as part of desrialization. given that, i don't think there is an approach that I can take to assert that fields are valid in a general sense until our runtime supports asking fields about their validity.

**Comment by habdelra** [comment]
@jurgenwerk the index card ordering comes from the query that the cards-grid component uses to get all the cards from the realm. Right now that query does not include any sorting--so it leaves the ordering up to the server. it looks like the server is coincidentally ordering search results in the order they were indexed when no sort order is supplied. since the auto save triggers indexing, the most recently saved card is the last card indexed, and hence appears at the end of the search results. I can file a ticket so that we can return cards in an order that is more stable by default--we'll need to decide  what that sort field should be though (probably `title` is a good place to start).

**Comment by habdelra** [comment]
actually the ordering of the cards grid is a super simple addition to the query. maybe I should just add title sorting in this PR...

**Comment by tintinthong** [comment]
Screenshot 2023-08-03 at 10 56 09 PM


Screenshot 2023-08-03 at 11 00 56 PM


The saving.. will be flooded by the titles if they r too long

**Review by tintinthong — APPROVED** [review]
Ive tested and it looks real good and robust. 

One thing I notice is creating empty contained cards will save empty contained cards. Particularly containsMany situations. A person can flood his card with empty contain cards from auto save. I dono if we want to consider anything about that

**Comment by habdelra** [comment]
> Screenshot 2023-08-03 at 10 56 09 PM
> 
> Screenshot 2023-08-03 at 11 00 56 PM
> 
> The saving.. will be flooded by the titles if they r too long

hmmmm, it looks like we need help with the design in this case. we can concatenate long titles but perhaps we'll need an official design for the saving message (it's just something that @lukemelia cooked up)

**Comment by habdelra** [comment]
> Ive tested and it looks real good and robust.
> 
> One thing I notice is creating empty contained cards will save empty contained cards. Particularly containsMany situations. A person can flood his card with empty contain cards from auto save. I dono if we want to consider anything about that

likely this will go away after the Card/Field refactor

**Comment by habdelra** [comment]
@jurgenwerk the card grid issues around ordering should be fixed now

File: packages/host/app/components/operator-mode/container.gts [hunk 2]
```diff
@@ -319,12 +304,38 @@ export default class OperatorModeContainer extends Component<Signature> {
...
+        format: Format = 'isolated',
+        fieldType?: 'linksTo' | 'contains' | 'containsMany' | 'linksToMany',
+        fieldName?: string
+      ) => {
+        let stack = here.stacks[stackIndex];
+        let itemsCount = stack.length;
+
+        let currentCardOnStack = here.getCard(stack[itemsCount - 1]!); // Last item on the stack
+
+        // TODO this is a hack until contained cards go away
```
Comment (backspace) [inline packages/host/app/components/operator-mode/container.gts:318]: is there somewhere to record this so it can be cleaned up when needed?

**Review by backspace — APPROVED** [review]

File: packages/host/app/components/operator-mode/container.gts (hunk 2 above)
Comment (habdelra) [reply packages/host/app/components/operator-mode/container.gts:318]: My hope is that the type change necessary will reveal all these places that need to update. Specially if the stack item type no longer includes `contained`, that should be enough for us to find all these items.

**Review by jurgenwerk — APPROVED** [review]
👍

**Pull request merged**